### PR TITLE
admin: fix recover expression cause index broken

### PR DESCRIPTION
**Pull request by xiongjiwei**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/41087 https://github.com/pingcap/tidb/issues/40430

Problem Summary:

### What is changed and how it works?

1. when rebuilding expression index in `recover index`, tidb will not calculate the virtual column and then write `nil` into the index
2. if the PK is int column and index contains PK, tidb with unistore will panic because the table scan part in the recover executor will have PK and `ExtraHandleID`, which cause unistore panic

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: executor/builder.go [hunk 1]
```diff
@@ -523,12 +523,29 @@ func (b *executorBuilder) buildRecoverIndex(v *plannercore.RecoverIndex) Executo
 		b.err = errors.Errorf("secondary index `%v` is not found in table `%v`", v.IndexName, v.Table.Name.O)
 		return nil
 	}
+	var containsGenedCol bool
+	cols := buildIdxColsConcatHandleCols(tblInfo, index.Meta(), func(indexInfo *model.IndexInfo) []*model.ColumnInfo {
+		columns := make([]*model.ColumnInfo, 0, len(indexInfo.Columns)+8)
+		for _, iCol := range indexInfo.Columns {
+			if tblInfo.Columns[iCol.Offset].IsGenerated() {
+				containsGenedCol = true
```
Comment (tangenta) [inline executor/builder.go:531]: Passing a function to `buildIdxColsConcatHandleCols` decrease the readability and make its name confusing.

I suggest extracting the initialization of `containsGenedCol` to a separate function, renaming `buildIdxColsConcatHandleCols` into `buildColsConcatHandleCols`, and passing a `hasVirtualCol bool` into it.

Comment (xiongjiwei) [reply executor/builder.go:531]: ok

Comment (xiongjiwei) [reply executor/builder.go]: done, @tangenta PTAL

**Review by tangenta — APPROVED** [review]

**Review by xhebox — APPROVED** [review]

**Pull request merged**